### PR TITLE
Fix initialization annotation of outer receiver with explicit annotation

### LINKED ISSUE #412: Type check receiver to forbid @UnknownInitialization receiver to be dereferenced
A mimimal test case could be:

```
    class InnerWithUnknownInitializationEnclosingExpression {
        InnerWithUnknownInitializationEnclosingExpression(
                @UnknownInitialization NullnessEnclosingExprTest NullnessEnclosingExprTest.this) {
            NullnessEnclosingExprTest.this.f.hashCode();
        }
    }
```

### PR DESCRIPTION
**Pull request by byd110**
### Problem description

fix #412. The explicit enclosing receiver annotation is captured by the receiver type of the constructor but not annotated in `InitializationParentAnnotatedTypeFactory.CommitmentTreeAnnotator#visitMethod`. 

This yeild a false positive during the check if we specify the enclosing receiver to be `@UnknownInitialized`, but in the method body check the type is the default `@Initialized`.

### Fix 
If the method visiting has an annotated receiver type, then retrive the annotation and annotate the enclosing receiver with it.

Two test cases are added to check this edge case. `checker/tests/initialization/Issue412a.java` checks the fix is correct. `framework/tests/h1h2checker/Issue412b.java` make sure the general framework can also catch this problem.

### CODE REVIEW

**Review by wmdietl — CHANGES_REQUESTED** [review]
Great, thanks for tracking this down and fixing this bug!
Please add an entry in the changelog, at least that the issue is fixed.

**Comment by wmdietl** [comment]
@byd110 Issue #412 mentions a possible relation to #226. Did you look into that?

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationParentAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -752,6 +752,21 @@ public Void visitMethod(MethodTree tree, AnnotatedTypeMirror p) {
                         (DeclaredType) exeType.getReturnType().getUnderlyingType();
                 AnnotationMirror a = getUnderInitializationAnnotationOfSuperType(underlyingType);
                 exeType.getReturnType().replaceAnnotation(a);
+
+                // If the receiver type exists (meaning there is an enclosing type) and is
+                // annotated, then annotate the enclosing type with the same annotation.
+                // TODO: look into why there is this inconsistency between receiver type and
```
[the remark this answers is not in the document]
Comment (byd110) [reply checker/src/main/java/org/checkerframework/checker/initialization/InitializationParentAnnotatedTypeFactory.java:758]: I would classify this as a saparate issue. `getEnclosingType` returns the enclsoing type with the correct annotation, whereas the enclosing type in the receiver is not annotated. 

This PR fully fixes the problem. The inconsistency only affacts the way to get the correct annotation.

I haven't looked into the root cause of this consistency. There is also no failing test cases to show this inconsistency would cause any problem.

**Comment by byd110** [comment]
> @byd110 Issue #412 mentions a possible relation to #226. Did you look into that?

I looked into it. It's another problem regarding method overide.

File: framework/tests/h1h2checker/Issue412b.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import org.checkerframework.framework.testchecker.h1h2checker.quals.*;
+
+// @skip-test
```
[the remark this answers is not in the document]
Comment (byd110) [reply framework/tests/h1h2checker/Issue412b.java:3]: Yes, it should be unskipped. The error reported rules out the problem with general framework.
